### PR TITLE
Fix loading status for Save attachments button

### PR DESCRIPTION
**Pull request by haivp3010**
Fix loading status for Save attachments button when closing file chooser without choosing. This fix is for both saving all attachments as well as saving a single attachment

### CODE REVIEW

**Comment by haivp3010** [comment]
#553

**Comment by ChristophWurst** [comment]
> First-time contributor

First of all, congratulations to your first PR to this repo and happy hacktoberfest 🎆 🚀 🎃

File: js/controller/messagecontroller.js [hunk 1]
```diff
@@ -120,13 +120,14 @@ define(function(require) {
 	 * @param {number} attachmentId
 	 * @returns {Promise}
 	 */
-	function saveAttachmentToFiles(account, folder, messageId, attachmentId) {
+	function saveAttachmentToFiles(account, folder, messageId, attachmentId, callback) {
```
Comment (ChristophWurst) [inline js/controller/messagecontroller.js]: @haivp3010 don't know whether you're familiar with the concept of promises, but this function returns one. This means we get an object that we can register callback on, which are executed as soon as the asynchronous task has finished. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises for more info 😉

File: js/views/messageattachment.js [hunk 2]
```diff
@@ -68,14 +68,14 @@ define(function(require) {
 			var folder = require('state').currentFolder;
 			var messageId = this.model.get('messageId');
 			var attachmentId = this.model.get('id');
-			// Loading feedback
-			this.getUI('saveToCloudButton').removeClass('icon-folder')
```
Comment (ChristophWurst) [inline js/views/messageattachment.js]: Why did you remove this? Doesn't this remove the initial loading feedback?

File: js/views/messageattachment.js [hunk 3]
```diff
@@ -68,14 +68,14 @@ define(function(require) {
...
-				.addClass('icon-loading-small')
-				.prop('disabled', true);
 
 			var _this = this;
-			MessageController.saveAttachmentToFiles(account, folder, messageId, attachmentId)
-				.catch(console.error.bind(this)).then(function() {
+			MessageController.saveAttachmentToFiles(account, folder, messageId, attachmentId, function() {
+				// Loading feedback
+				_this.getUI('saveToCloudButton').removeClass('icon-folder')
```
Comment (ChristophWurst) [inline js/views/messageattachment.js]: see five lines below, the code is already there 😉 For some reason it doesn't seem to work though. Try to debug and fix that instead please

File: js/views/messageattachments.js [hunk 4]
```diff
@@ -57,14 +57,14 @@ define(function(require) {
...
-				.addClass('icon-loading-small')
-				.prop('disabled', true);
 
 			var _this = this;
-			MessageController.saveAttachmentsToFiles(account, folder, messageId)
-				.catch(console.error.bind(this)).then(function() {
+			MessageController.saveAttachmentsToFiles(account, folder, messageId, function() {
+				// Loading feedback
+				_this.getUI('saveAllToCloud').removeClass('icon-folder')
+				.addClass('icon-loading-small')
```
Comment (ChristophWurst) [inline js/views/messageattachments.js]: same as above, you can find the promise callback a few lines below

File: js/controller/messagecontroller.js (hunk 1 above)
Comment (haivp3010) [reply js/controller/messagecontroller.js]: I do know about promises, but in this case, I just haven't found out if that close event of the file chooser calls the reject method of the promise.

File: js/views/messageattachment.js (hunk 2 above)
Comment (haivp3010) [reply js/views/messageattachment.js]: I wanted to make the button show loading only when user clicks Choose to choose a directory. Shouldn't the button stay the same unless user confirms saving?

File: js/views/messageattachment.js (hunk 3 above)
Comment (haivp3010) [reply js/views/messageattachment.js]: I moved it here to only show loading when user confirms directory, as above.

**Comment by haivp3010** [comment]
Thanks, I am looking into that now :)

File: js/controller/messagecontroller.js (hunk 1 above)
Comment (ChristophWurst) [reply js/controller/messagecontroller.js]: I don't know that either as I haven't checked. But from the diff here on GitHub I can tell that you're invoking the callback as soon as the user selected a directory. However, the loading feedback should be there until the task of actually storing the attachments has been finished.

Comment (ChristophWurst) [reply js/controller/messagecontroller.js]: Wait. Ignore my comment, now I see what you did there 🙈 The callback was added for *adding* the feeback, not for removing.

**Comment by ChristophWurst** [comment]
> Thanks, I am looking into that now :)

Might be that you're actually correct with your patch. I will check the code soon :)

File: js/controller/messagecontroller.js (hunk 1 above)
Comment (haivp3010) [reply js/controller/messagecontroller.js]: Yeah. That's what I meant: no loading until user has chosen a directory, and loading disappears after upload done. I have checked that on my localhost and it went fine. It's just that I couldn't find a way to achieve this with promise, with the existing code. I had thought of catching the close event on the file chooser and call reject, but that would require making changes to core js files of nextcloud server as well, so I had to resort to callback like that.

**Comment by haivp3010** [comment]
Thanks. I have just pushed a small patch for checking the callback. Please check back when you have time :)

**Comment by haivp3010** [comment]
Any progress with this yet? @ChristophWurst

**Comment by ChristophWurst** [comment]
> Any progress with this yet? @ChristophWurst

No, because I'm busy with other stuff, sorry. Can anybody of @nextcloud/mail help here please? Thanks.

File: js/views/messageattachment.js (hunk 2 above)
Comment (jancborchardt) [reply js/views/messageattachment.js]: @haivp3010 the loading feedback should appear as soon as the button is clicked already. So if the popover does _not_ overlap the button you see the loading feedback already.

Comment (haivp3010) [reply js/views/messageattachment.js]: @jancborchardt I think that the purpose of the loader is to represent file saving status. For the file picker, it does have its own loading spinner while loading up, so why should we leave the former as loading while it's function is actually not being executed?

**Comment by haivp3010** [comment]
It's ok, I was just kind of eager :) Just give it a look later when you have time. Thanks.

File: js/views/messageattachment.js (hunk 2 above)
Comment (haivp3010) [reply js/views/messageattachment.js]: A similar situation can be seen here https://developers.google.com/drive/v3/web/savetodrive, where the Save to drive button does not show as loading while still choosing a folder.
image
The loader only shows when the file is actually being uploaded.
image

Of course, different systems have different ways to do things, but I think it's just not common practice to show the loading feedback so early while still choosing a directory.
